### PR TITLE
feat: standalone HTML view with table highlighting

### PR DESCRIPTION
**Pull request by jhpedemonte**
#### What do these changes do/fix?

Enables `HtmlView` to with table highlighting.

Screen Shot 2020-01-12 at 6 59 46 PM

https://github.ibm.com/Watson-Discovery/docviz-squad-issue-tracker/issues/220
https://github.ibm.com/Watson-Discovery/docviz-squad-issue-tracker/issues/221

#### How do you test/verify these changes?

Storybook > DocumentPreview > components > HtmlView

### CODE REVIEW

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx [hunk 1]
```diff
@@ -18,21 +26,115 @@ interface Props {
...
+          </div>`);
+        };
+        process();
+      } else {
+        // if no highlight, then no need to "process"
+        setHtml(docHtml ? DOMPurify.sanitize(docHtml, SANITIZE_CONFIG) : '');
+      }
+    }
+
+    setLoading(false);
```
Comment (maniax89) [inline packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: don't you want to wait until the above `process` command is complete to set loading to false? im not sure how long  `processDoc` takes but seems that you might want to wait to `setLoading` until after that is done

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx [hunk 2]
```diff
@@ -18,21 +26,115 @@ interface Props {
...
+  }, [document, highlight, setLoading]);
```
Comment (maniax89) [inline packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: i think the `useEffect` here will re-run a bunch for these non-primitive, non-react-generated input parameters. not sure if that is a concern for performance as I haven't tested it locally. one option is to use the `useDeepCompareEffect` which uses lodash's `isEqual` method on complex object types

regardless, this is a performance optimization only to revisit if we see things being slow, etc.

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx [hunk 3]
```diff
@@ -18,21 +26,115 @@ interface Props {
...
+
+    if (highlightNode) {
+      clearNodeChildren(highlightNode);
+    }
+
+    if (!highlight || !contentNode || !highlightNode) {
+      return;
+    }
+
+    const { begin, end } = get(highlight, 'table.location');
```
Comment (maniax89) [inline packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: should the fallback be `{}` here?

```suggestion
    const { begin, end } = get(highlight, 'table.location', {});
```

worried that trying a variable destructuring assignment will result in an error on `undefined`

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx [hunk 4]
```diff
@@ -18,21 +26,115 @@ interface Props {
...
+    });
+    highlightNode.appendChild(fragment);
 
-  setLoading(false);
+    // scroll highlight into view
+    const firstFieldRect = highlightNode.querySelector('.field--rect');
+    if (firstFieldRect) {
+      firstFieldRect.scrollIntoView({ block: 'center' });
+    }
+  }, [html, highlight]);
```
Comment (maniax89) [inline packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: same comment above about `useDeepCompareEffect`

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx [hunk 5]
```diff
@@ -18,21 +26,115 @@ interface Props {
...
+    'page'
+  ],
+  WHOLE_DOCUMENT: true
+};
+
+const base = `${settings.prefix}--html`;
+
+export const HtmlView: FC<Props> = ({
+  document,
+  highlight,
```
Comment (maniax89) [inline packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: what is `highlight` in this context? is it the `table_results` data?

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx (hunk 1 above)
Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: Good catch.

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx (hunk 2 above)
Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: So far, haven't seen perf issues.

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx (hunk 3 above)
Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: Another good catch!

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx (hunk 4 above)
Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: Same answer.

File: packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx (hunk 5 above)
Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/components/HtmlView/HtmlView.tsx]: Yes.

**Pull request merged**